### ISSUE TITLE
NumberControl tweaker buttons should be able to match the size of the NumberDisplay

### ISSUE THREAD

**Issue by samreid** [phetsims/scenery-phet#513]
In https://github.com/phetsims/scenery-phet/issues/508#issuecomment-507062895 I said:

> Designers have requested the tweaker buttons height to match the height of the number readout display, but the NumberControl API for the tweaker buttons is via arrowButtonOptions: {scale}, so the current implementation is jury-rigged. Is there a more solid way to do it?

@pixelzoom replied:

> Something like this in `NumberControl`:
```js
// By default, scale the ArrowButtons to have the same height as the NumberDisplay.
if ( options.arrowButtonOptions.scale === undefined ) {
  const arrowButtonsScale = numberDisplay.height / leftArrowButton.height; 
  leftArrowButton.setScaleMagnitude( arrowButtonsScale );
  rightArrowButton.arrowButtonsScale( arrowButtonScale );
}
```

UPDATE: Tagging for https://github.com/phetsims/wave-interference/issues/408

**Comment by samreid** [comment]
Would it be better to use `null` instead of `undefined` so we don't end up passing `undefined` from client code?  Would we change the default value from `0.85` to `null`?  And, if so, will we need to check through all sims that used the old default to make sure the new default is acceptable?

**Comment by pixelzoom** [comment]
> Would it be better to use `null` instead of `undefined`?

Since `scale` is `Node` option, I would probably use whatever scenery used for the default. Looking at `DEFAULT_OPTIONS` in `Node`, it looks like the default for `scale` is probably `undefined`. 

> Would we change the default value from `0.85` to `null`?  

Yes.  Unless it's OK to push this new convention into all sims, which is a question for designers.

> And, if so, will we need to check through all sims that used the old default to make sure the new default is acceptable?

Yes.

**Comment by samreid** [comment]
I tried using this code in Wave Interference:
```js
arrowButtonOptions: {
  scale: undefined
},
```

And was met with 

> Assertion failed: Undefined not allowed for Node key: scale

Therefore it does not seem there is a simple way for pre-existing simulations to have `0.85` and other occurrences opt-in to undefined.

@pixelzoom how do you recommend to proceed?

**Comment by pixelzoom** [comment]
There's no reason to set `scale: undefined` explicitly as the default in `arrowButtonOptions`.   You want to test for whether the caller provided this option.   If testing for `options.arrowButtonOptions.scale === undefined` seem distasteful, then use `!options.arrowButtonOptions.hasOwnProperty( 'scale' )`.

**Comment by samreid** [comment]
Here's a change set that defaults the arrow buttons to the height of the number display:

<details>

```diff

Index: wave-interference/js/common/WaveInterferenceConstants.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- wave-interference/js/common/WaveInterferenceConstants.js	(revision 4688cc1aa0cf83de79dc1f3c49d8e6fd53590393)
+++ wave-interference/js/common/WaveInterferenceConstants.js	(date 1563227871000)
@@ -44,7 +44,6 @@
     NUMBER_CONTROL_OPTIONS: {
       layoutFunction: NumberControl.createLayoutFunction4( { verticalSpacing: 3 } ),
       arrowButtonOptions: {
-        scale: 0.65,
         touchAreaXDilation: 9,
         touchAreaYDilation: 10
       },
Index: scenery-phet/js/WavelengthNumberControl.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- scenery-phet/js/WavelengthNumberControl.js	(revision de648f0e5ac55f8a0e44e9522044706fbce0b7ae)
+++ scenery-phet/js/WavelengthNumberControl.js	(date 1563227795000)
@@ -45,9 +45,6 @@
       const trackHeight = options.trackHeight;
 
       super( options.title, property, options.range, merge( {
-        arrowButtonOptions: {
-          scale: trackHeight * 0.0315 // roughly the height of the track
-        },
         titleNodeOptions: {
           font: new PhetFont( 15 ),
           maxWidth: 175
Index: scenery-phet/js/NumberControl.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- scenery-phet/js/NumberControl.js	(revision de648f0e5ac55f8a0e44e9522044706fbce0b7ae)
+++ scenery-phet/js/NumberControl.js	(date 1563228764000)
@@ -19,6 +19,7 @@
   const HBox = require( 'SCENERY/nodes/HBox' );
   const HSlider = require( 'SUN/HSlider' );
   const inherit = require( 'PHET_CORE/inherit' );
+  const InstanceRegistry = require( 'PHET_CORE/documentation/InstanceRegistry' );
   const merge = require( 'PHET_CORE/merge' );
   const Node = require( 'SCENERY/nodes/Node' );
   const NumberControlIO = require( 'SCENERY_PHET/NumberControlIO' );
@@ -88,12 +89,13 @@
       groupFocusHighlight: true
     }, options );
 
+    const arrowButtonScaleProvided = options.arrowButtonOptions && options.arrowButtonOptions.hasOwnProperty( 'scale' );
+
     // Merge all nested options in one block.
     options = merge( {
 
       // Options propagated to ArrowButton
       arrowButtonOptions: {
-        scale: 0.85,
 
         // Values chosen to match previous behavior, see https://github.com/phetsims/scenery-phet/issues/489.
         // touchAreaXDilation is 1/2 of its original value because touchArea is shifted.
@@ -243,6 +245,16 @@
       tandem: options.tandem.createTandem( 'rightArrowButton' )
     }, options.arrowButtonOptions ) );
 
+    // By default, scale the ArrowButtons to have the same height as the NumberDisplay.
+    if ( !arrowButtonScaleProvided ) {
+
+      leftArrowButton.setScaleMagnitude( 1 );
+      const arrowButtonsScale = numberDisplay.height / leftArrowButton.height;
+
+      leftArrowButton.setScaleMagnitude( arrowButtonsScale );
+      rightArrowButton.setScaleMagnitude( arrowButtonsScale );
+    }
+
     // arrow button touchAreas, asymmetrical, see https://github.com/phetsims/scenery-phet/issues/489
     leftArrowButton.touchArea = leftArrowButton.localBounds
       .dilatedXY( arrowButtonPointerAreaOptions.touchAreaXDilation, arrowButtonPointerAreaOptions.touchAreaYDilation )
@@ -333,6 +345,9 @@
       numberProperty.unlink( arrowEnabledListener );
       this.enabledProperty.unlink( enabledObserver );
     };
+
+    // support for binder documentation, stripped out in builds and only runs when ?binder is specified
+    assert && phet.chipper.queryParameters.binder && InstanceRegistry.registerDataURL( 'scenery-phet', 'NumberControl', this );
   }
 
   /**

```

</details>

We cannot have some sims use one default and other sims use another default, so I think we will need to see if we can make this change for all simulations.  Thanks to binder, here is a list of sims that use NumberControl:

beers-law-lab
coulombs-law
energy-skate-park
gas-properties
gravity-force-lab
hookes-law
masses-and-springs
masses-and-springs-basics
pendulum-lab
plinko-probability
projectile-motion
rutherford-scattering
wave-interference

I'll check through them to see how different the default looks.

**Comment by samreid** [comment]
When I restricted the binder search to sims that use the defaults for NumberControl, it returned this list:

gas-properties
hookes-law
plinko-probability
rutherford-scattering
wave-interference

# gas properties
the change looks minor, but seems acceptable to me
Original
![image](https://user-images.githubusercontent.com/679486/61255954-05d6bb80-a728-11e9-8699-e550874777f8.png)
Proposed change
![image](https://user-images.githubusercontent.com/679486/61255961-0ff8ba00-a728-11e9-9b4c-f1835da158ec.png)

# hookes law
tweaker buttons become slightly smaller with this change
Original
![image](https://user-images.githubusercontent.com/679486/61256014-564e1900-a728-11e9-9eb9-c08f49490654.png)
Proposed
![image](https://user-images.githubusercontent.com/679486/61256036-69f97f80-a728-11e9-9b8c-51029b47067e.png)

# plinko probability
Original
![image](https://user-images.githubusercontent.com/679486/61256085-944b3d00-a728-11e9-80ed-c983052d620a.png)
Proposed
![image](https://user-images.githubusercontent.com/679486/61256088-9b724b00-a728-11e9-880c-ed8cde68af21.png)

# rutherford scattering
Original
![image](https://user-images.githubusercontent.com/679486/61256294-7fbb7480-a729-11e9-858e-56410639bca8.png)

Proposed
![image](https://user-images.githubusercontent.com/679486/61256310-877b1900-a729-11e9-947c-e7545979cb20.png)

And wave interference was requested to have the tweaker buttons have the same height as the number display, so that's a safe change.

@arouinfar do these changes seem acceptable to you?  Do we need other designers to sign off before I proceed?

**Comment by pixelzoom** [comment]
Couldn't help lurking and throwing my opinion in :)  They all look mighty fine to me too.   Might be good to also check with `?showPointerAreas` to make sure there's nothing whacky going on there.

**Comment by samreid** [comment]
I tested hookes law, and it looks like the pointer areas changed proportionately:

Original
![image](https://user-images.githubusercontent.com/679486/61258765-78e62f00-a734-11e9-8c83-dc1032bd41c5.png)

Proposed
![image](https://user-images.githubusercontent.com/679486/61258771-800d3d00-a734-11e9-8604-b3fa793468e7.png)

**Comment by arouinfar** [comment]
Looks great @samreid! It's a subtle change, but I think matching tweaker height to the readout makes things look cleaner overall.

**Comment by samreid** [comment]
Changes pushed, @pixelzoom can you please review?

**Comment by pixelzoom** [comment]
Changes reviewed, looks good.  Feel free to close if work is completed.

**Comment by samreid** [comment]
All set, thanks, closing.

**Comment by samreid** [comment]
In https://github.com/phetsims/wave-interference/issues/420 @KatieWoe pointed out a flaw with this design rule.

We set the tweaker button size to match the size of the number display. Lengthy text can shrink the number display, so the arrow buttons have shrunken too. This problem will impact every sim with NumberControl.

![image](https://user-images.githubusercontent.com/679486/62408133-e097d900-b580-11e9-9e5a-9dc24e64f6d5.png)

It seems one way around this for Wave Interference would be to specify the `arrowButtonOptions.scale` instead of letting it default to the NumberDisplay height for this case.  But I'm skeptical that this is a scalable solution--it would be unfortunate if we have to apply that pattern across numerous sims.  I've also considered solutions like only matching the tweaker button height to the NumberDisplay height if they appear adjacent to each other, or limiting the amount the tweaker buttons can grow or shrink.  Or making it so the tweaker buttons follow the un-scaled height of the text box (before it runs into maxWidth problems)?  I'm really grasping at straws here.

@pixelzoom what do you recommend?

**Comment by pixelzoom** [comment]
>  I'm really grasping at straws here.

Me too, no good solution comes to mind. I will ruminate.

**Comment by pixelzoom** [comment]
Rumination complete :)  

My recommendation is to back out this change, and require the client to adjust the size of the arrow buttons.   If the changes that @arouinfar identified as "cleaner overall" in https://github.com/phetsims/scenery-phet/issues/513#issuecomment-511857309 are worth keeping, note the arrow button sizes before reverting, and set them explicitly after reverting.

**Comment by pixelzoom** [comment]
More...

I'm not sure that reverting is the best option, but it's the best I could come up with.  Setting arrow button sizes explicitly will result in fixes sizes that won't adapt to change in NumberDisplay size that occur due to font size differences across platforms.  But this never bothered me previously, seemed to work fine. And the improvement was described by @arouinfar as "subtle", so maybe not necessary.

On the other hand...  The example shown in https://github.com/phetsims/scenery-phet/issues/513#issuecomment-517897850 is highly unlikely to occur. So while I appreciate @KatieWoe's attention to detail in discovering this, perhaps we could just leave "as is" and everything would be fine in practice.

**Comment by pixelzoom** [comment]
Still more...

So the options that I can come up with are:

(1) Revert and live with it.

(2) Leave as is and live with it.

**Comment by samreid** [comment]
One more part of this problem.  I don't see a clear way to explicitly set the height of these buttons, the API is instead centered around `arrowButtonOptions.scale`.  The default value is `1`, but it's unclear how big that is exactly, or what it is supposed to match.  I'm seeing several sims that are using `0.56` or `0.5` or `0.55` for the scale.  Should this API ultimately be changed to be about width and/or height?  Or possibly just one length dimension (and assuming it is square or nearly square or keeps its aspect ratio)?  This could be done in a separate issue but it seems like it equally applies to the cases of (a) we revert changes and do not try to match the NumberDisplay height or (b) a sim needs to opt out of matching the NumberDisplay height by passing an option.

**Comment by samreid** [comment]
@arouinfar can you please review https://github.com/phetsims/scenery-phet/issues/513#issuecomment-517897850 and below?  What do you recommend?

**Comment by samreid** [comment]
Another solution would be to (by default) set the height of the tweaker buttons to match the height of the unscaled `numberDisplay`.  @jonathanolson pointed out this can be done by using the `localBounds.height` of the number instead of just the `height`.  This appears OK in local testing and resolves the shrinking button problem, I'll commit and run this past @pixelzoom.

**Comment by samreid** [comment]
@pixelzoom can you please review?

**Comment by pixelzoom** [comment]
This looks good.  I'd like to see a comment in the code about why we're ignoring NumberDisplay `maxWidth` and instead using `localBounds`, specifically mentioning the problem reported in https://github.com/phetsims/scenery-phet/issues/513#issuecomment-517897850 (a link to that comment would be good too).

**Comment by samreid** [comment]
I added docs above.  It felt a bit wordy but I wasn't sure how to improve it.  @pixelzoom can you please review?

**Comment by pixelzoom** [comment]
👍 Looks good to me, not too wordy.  Closing.
